llama_index/callbacks/base.py

```diff
@@ -88,3 +113,36 @@ def remove_handler(self, handler: BaseCallbackHandler) -> None:
     def set_handlers(self, handlers: List[BaseCallbackHandler]) -> None:
         """Set handlers as the only handlers on the callback manager."""
         self.handlers = handlers
+
+    @contextmanager
+    def as_trace(self, run_id: str) -> Generator[None, None, None]:
+        """Context manager tracer for lanching and shutdown of runs."""
+        self.launch(run_id=run_id)
+        yield
+        self.shutdown(run_id=run_id)
+
+    def launch(self, run_id: Optional[str] = None) -> None:
+        """Run when an overall run is launched."""
+        if not self._running_id:
+            self._trace_map = defaultdict(list)
+            self._trace_stack = [BASE_TRACE_ID]
+
+            for handler in self.handlers:
+                handler.launch(run_id=run_id)
+
+            self._running_id = run_id
+
+    def shutdown(
+        self,
+        run_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        """Run when an overall run is exited."""
+        if run_id is not None and run_id == self._running_id:
+            for handler in self.handlers:
+                handler.shutdown(run_id=run_id, trace_map=self._trace_map)
+            self._running_id = None

```

Comment:
it's not super clear to me whether we should reset trace_map and trace_stack during launch, shutdown, or both

(out of curiosity do we need both methods?)